chain/chain/src/chain.rs

```diff
@@ -4714,22 +4737,24 @@ impl Chain {
         self.invalid_blocks.contains(hash)
     }
 
-    /// Check if can sync with sync_hash
+    /// Check that sync_hash is the frst block of an epoch.
     pub fn check_sync_hash_validity(&self, sync_hash: &CryptoHash) -> Result<bool, Error> {
-        let head = self.head()?;
         // It's important to check that Block exists because we will sync with it.
-        // Do not replace with `get_block_header`.
+        // Do not replace with `get_block_header()`.
         let sync_block = self.get_block(sync_hash)?;
-        // The Epoch of sync_hash may be either the current one or the previous one
-        if head.epoch_id == *sync_block.header().epoch_id()
-            || head.epoch_id == *sync_block.header().next_epoch_id()
-        {
-            let prev_hash = *sync_block.header().prev_hash();
-            // If sync_hash is not on the Epoch boundary, it's malicious behavior
-            Ok(self.epoch_manager.is_next_block_epoch_start(&prev_hash)?)
-        } else {
-            Ok(false) // invalid Epoch of sync_hash, possible malicious behavior
-        }
+        let prev_hash = *sync_block.header().prev_hash();
+        let is_first_block_of_epoch = self.epoch_manager.is_next_block_epoch_start(&prev_hash);
```

Comment:
That check is replaced by checking whether the node has a matching state snapshot.
If a node has no snapshot, then no sync_hashes are accepted.
If a node happens to have an old snapshot, then it will accept those old sync hashes.

Reply:
If a node has no sync hashes it should not reply because probably did not advertise that it can share parts. 
What happens if a node tracks shard 0 and the snapshot has parts only for that shard but the node gets a request for shard 3? How will the reply look like? 
I see it something like:
```
{
  part: None,
  can_generate: false,
  bitmap: Some(Empty) // or None
}